auth-add-authPage

arifm6 opened the pull request:
# Changes
### 1. Added logic for signin, signup, and recover-account page.

- created auth forms to handle dynamic state changes. 
- separated auth forms and page containers into reusable components. The previous implementation was too cluttered and not reusable. 
- set up back end logic for submitting signin form, signup form, and recover-account form.

### 2. updated db schema and set up email api.
- updated db schema
- set up email api key and sending service to send emails from backend to user email

### 3. minor changes to other files. 
- added barebones secondary auth pages for additional tasks. 
- created and temporarly disabled hooks for input validation.
- created util files for necessary logic.
- created trpc routes and query and mutation logic for CRUD operations on the db.
- other minor changes not listed above.


**Notion task** - https://www.notion.so/39d5f132e6dd483ea6a128c7152f5009?v=159487785be744a39ac2478076bbe410&p=54ce9a78fd51436eb1e7ae1c5c9042b1&pm=s

This task done but will not be able to test without setting up the necessary env files and external providers. I will provide documentation on how to set this up and will leave the PR as draft in the meantime. 







<br/>
<br/>

# Roadmap


- [ ] 1 - Don't use negative conditions

- [ ] 2 - Simplify code

- [ ] 3 - Use supabase tables

- [x] 4 - Delete native comments everywhere

- [ ] 5 - Delete .txt files

- [ ] 6 - ~~Change file extention to .ts~~

- [ ] 7 - Fix eslint errors

- [ ] 8 - Don't use 'as string' with keys

- [ ] 9 - ~~Don't use negative conditions~~

- [ ] 10 - Use modern emails please

- [ ] 11 - ~~Change ctx to context~~

- [x] 12 - Use figma design (naming)

- [ ] 13 - Use figma design (icons)

- [ ] 14 - Please write much better 'Changes' next time

- [x] 15 - Use components instead of native button tags

- [ ] 16 - Decide which input validation do you use

- [ ] 17 - Improve UX (validation errors)

- [ ] 18 - Let user undestand somehow that he is logged in

- [ ] 19 - ~~Fix issue when I'm still see 'login' when I logged in~~

- [ ] 20 - Use cookie-based auth

- [ ] 21. Create the error page

- [ ] 22. Specify id in auth.ts

- [ ] 23. Improve 'Changes'

arifm6 committed:
separated auth form logic into reusable components and tested for signin page

arifm6 committed:
updated input component. Added simple checkbox component. made auth pages more reusable by splitting into components. updated ProductsHeader component to reflect new input component change

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
Update AuthContainer.tsx

changed max width from 80vw to 90vw.

arifm6 committed:
updated forms to handle real time change using react state. Updated input field (password) to focus on visibility change

arifm6 committed:
Minor Updates

updated prisma to version 5, schema, and cleaned up eslint warnings/errors

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
Minor import path change

Fixed import path which was causing an error and fixed ESLint warnings.

arifm6 committed:
Update index.ts

Commit:
z deleted

arifm6 committed:
split input components to specific types

arifm6 committed:
Merge branch 'auth-add-authPage' of https://github.com/Nicitaa/22-aer into auth-add-authPage

arifm6 committed:
fixed inputs and auth forms bugs

arifm6 committed:
updating code base to reflect new ui inputs

arifm6 committed:
fixed ESLint errors

arifm6 committed:
deleted SimpleCheckbox component

arifm6 committed:
Removed bracket strings

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
merged development into auth-add-authPage

arifm6 committed:
deleted inputs and resolved minor errors

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
Set up email signup validation

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
added auth validation and email check on signup. Fixed errors after merge from development

arifm6 committed:
added functional signup page with error handling.

arifm6 committed:
added basic auth page functionality. Register and user login available with no error handling or email verification/recover account

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
generate verification token on user signup for email verification.

arifm6 committed:
set up email verification with send grid. Updated prisma schema.

arifm6 committed:
finished auth pages with email sending functionality. Updated schema to include recoveryToken. updated pages to handle redirect better. adding barebones new pages that are required and necessary API endpoints. commented out previous validation hooks temporarly until improved.

arifm6 committed:
Merge branch 'development' into auth-add-authPage

arifm6 committed:
removed continue with credentials button

nicitaacom commented:
I find its good that you do spearate commit for each task - it helps me understand what you did in every commit (it is not necessary but preffered)

I find its bad that you do long commit names - plase use shrot commit names that shortly described all context
![image](https://github.com/nicitaacom/22_aer/assets/39565703/83b1a89c-db5b-45be-8b95-e0cc97452972)

nicitaacom commented:
About this commit 
![image](https://github.com/nicitaacom/22_aer/assets/39565703/ddfac962-6104-47dd-b245-a577193b1fbc)

I didn't understood what you merged here
If you update your branch from development - name it simply 'update from development (#nn)

nicitaacom commented:
I just look through these changes - in some points (like input validantion and error hadling it looks complicated)

Especially here
![image](https://github.com/nicitaacom/22_aer/assets/39565703/745d8fa7-9c64-4564-a289-f77788815ddd)

Please use clean code rule that says 'don't use nagative conditions' - it means that you don't need use smth like !email ro !emailMessage
emailMessage or validatePasswordMessage tells me nothing (some message about password validation but it from sever or it local input validation? - unclear)

Also a lot of nested conditions like if password.length else if (...) return
You can simplify it like
```tsx
if (password.length) {
//do stuff
}
return ""
```
Also even this return "" I find unnecessary because by default you have <string | undefined> (or default message like 'email should be valid' that hidden by default)

I would reccomend you to take a look of videos on YouTube about 'clean code' and 'maintainable code'

Keep in mind in feature when project comes to end or you will work for another company somebody else who will work on your position must understand what's going on in their part code what he/she responsive for
(e.g I responsive for UI and now problem is nobody understand what's going on in UI (only basic understanding) - and solution is do some meet ups and speek about something we want to speek for example this points I described above)

**Note**:This is important and I would recommend you if you don't get my idead - read it one more time and/or ask a question

nicitaacom commented:
Also about notion - we almost use of our free 1000 block 
Here is 2 solutons that I see:
1.We can buy workspace for ~240$ (I find it bad because I not able to pay so much - you know my internet)
2. Adaptive to github issues (github have similar functionalit and it also helps us to keep everyting in 1 workspace)

I'd like to use second solution because it free

nicitaacom commented:
Also 1 thing about notion
Please don't create empty pages like this

### Use methodology
Please describe problem that we have and solution that solves the problem
Not only this case

![image](https://github.com/nicitaacom/22_aer/assets/39565703/6afd9ab0-460a-442b-a4e0-7bf47fa07714)

arifm6 committed:
combined validation functions and messages

nicitaacom commented:
Please change any possible way file extention from .mjs to .ts

arifm6 committed:
Updated validation utility functions

updated 3 validation functions in utility functions and removed unused code in signin and signup forms.

Commit:
Merge branch 'development' into auth-add-authPage

Commit:
Merge branch 'auth-add-authPage' of https://github.com/nicitaacom/22_aer into auth-add-authPage

nicitaacom commented:
# Requested changes_2

<details>
<summary>1. Don't use negative conditions</summary>

![image](https://github.com/nicitaacom/22_aer/assets/39565703/5d1a3e90-c279-493e-9f90-10f79b942cb6)
Use .length > 0 or .length === 0 or
```ts
 if (password) {
return stuff
}
return stuff
```

</details>




<details>
<summary>2. Simplify code</summary>
Now your code looks complicated - please simplify code and write docs for your part

When somebody else will work on this porject (in auth part) he need understand what's going on in your code

</details>






<details>
<summary>3. Use supabase tables</summary>

We use supabase (supabase already has PostgreSQL)

It is unconsistant to create custom .sql tables  (In mean directly in project instead of supabase tables)

![image](https://github.com/nicitaacom/22_aer/assets/39565703/ff11aab7-a9ce-4a42-9f7b-255c3fb42a58)

Please use supabase tables 
</details>








<details>
<summary>4. Delete native comments everywhere</summary>
Please delete this comments

![image](https://github.com/nicitaacom/22_aer/assets/39565703/5644d4c0-5ab2-42b4-b62b-a5afd4b98ce9)

![image](https://github.com/nicitaacom/22_aer/assets/39565703/a1ab35cb-d8b7-4a35-a0cb-e1359f490710)

![image](https://github.com/nicitaacom/22_aer/assets/39565703/1869bea2-4bae-4839-8113-4084e3056541)


It helps to keep code clean
</details>






<details>
<summary>5.Delete .txt files</summary>
Sometimes I create .txt file as well (to keep something in clipboard)

But when you commit changes and mark PR as 'ready to review' - please delete .txt files
![image](https://github.com/nicitaacom/22_aer/assets/39565703/63577cb4-65c9-4759-a6db-c39e11a2b93e)

</details>






<details>
<summary>6. Change file extention to .ts</summary>

![image](https://github.com/nicitaacom/22_aer/assets/39565703/08461d23-94bf-4999-b3a3-8cee5c296069)

We use TypeScript - change file extention any possible way
</details>







<details>
<summary>7. Fix eslint errors</summary>

![image](https://github.com/nicitaacom/22_aer/assets/39565703/c599324b-6e1c-454c-af25-f13f53966dfb)

We use TypeScript - so please specity types
If you use library install necessary types with `pnpm i @types/package-name`

</details>







<details>
<summary>8. Don't use 'as string' with keys</summary>

![image](https://github.com/nicitaacom/22_aer/assets/39565703/9c79a8f5-0f8d-4f06-abe5-f6da0fc33c64)

If you want to say TypeScript that this value is 'non-null' use ! at the end like
`process.env.NEXT_SENDGRID_PUBLIC!`

</details>







<details>
<summary>9. Don't use negative conditions</summary>
To keep code consistant and undestandable please don't use negative condition

![image](https://github.com/nicitaacom/22_aer/assets/39565703/9e70fee6-2200-4cc8-99cf-3830dcdf37d8)

Use smth like instead
```
if (user) {
return stuff
}

return "You need login to perform this action"
```
So delete all negative conditions in your code

</details>







<details>
<summary>10. Use modern emails please</summary>
You can use html here - so be creative and create cool recovery email

![image](https://github.com/nicitaacom/22_aer/assets/39565703/9d1a22eb-5868-4470-8b82-e2755e1ef631)


Find impression you can by clicking 'forgot password' on some websites and chech how email looks like

</details>





<details>
<summary>11. Change ctx to context</summary>

Somethig that not so important but I don't like it

I know that its 'my wish' or somthing like that but it would be better when you name your variables like this
`context` instead of `ctx`

![image](https://github.com/nicitaacom/22_aer/assets/39565703/a0a1b835-db92-4450-8c64-b18d93914552)

(In my head it something like - we use tailwind and tailwind names classes like `flex justify-center` etc instead of `flx jcc` etc)
(also I see that you write back end - tbh nobody cares but it would be better when you will write backend to be consistant)
So be consistant here 

</details>







<details>
<summary>12. Use figma design (naming)</summary>
It names footer_login_help (or somthing like this I don't remember exactly)

But defentely not customContent

![image](https://github.com/nicitaacom/22_aer/assets/39565703/3beae5c1-7165-4c24-ab0d-b99c92c0d131)


In figma we have names of groups - so use this names
(and please decide you use CamelCase *or* snake_case *or* kebab-case) - be consistant in this point
</details>







<details>
<summary>13. Use figma design (icons)</summary>
We have another icons in figma design

![image](https://github.com/nicitaacom/22_aer/assets/39565703/6915b3b1-e5f9-4ef0-b44d-aa80f9754db6)


We use outline style for icons

</details>






<details>
<summary>14. Please write much better 'Changes' next time</summary>
For example

> updated schema db

tells me nothing and may ommited
![image](https://github.com/nicitaacom/22_aer/assets/39565703/ac04fd7a-895a-4dde-8aa8-0ba8cfcbcd9f)

Write only necessary changes that solves some problems
Also describle which problem you solved with your solution
(e.g created recover-account pages - now user able to recover account using this page)

</details>







<details>
<summary>15. Use components instead of native button tags</summary>
![image](https://github.com/nicitaacom/22_aer/assets/39565703/7115dc9f-3c19-462b-9cc8-53ab59b0ac18)

Also change it not here but in all files where smb use <button> tag
</details>







<details>
<summary>16. Decide which input validation do you use</summary>
![image](https://github.com/nicitaacom/22_aer/assets/39565703/1e817d71-8063-444d-b468-6ab5fc1b79e5)

Please disable this native input validation messages
(you may use react-hook-form)
</details>







<details>
<summary>17. Improve UX (validation errros)</summary>
Really?

![image](https://github.com/nicitaacom/22_aer/assets/39565703/3f38eb80-3205-49a9-83bf-5a513470e904)

</details>





<details>
<summary>18. Let user undestand somehow that he is logged in</summary>

![image](https://github.com/nicitaacom/22_aer/assets/39565703/cd896761-8338-444e-a9af-696b4c90905b)


You have great opportunity to upgrady your figma skills and create figma design for user authentication (at least you can display message like 'you logged in' or something like this
</details>





<details>
<summary>19. Fix issue when I'm still see 'login' when I logged in</summary>
make `display:hidden` this nav-link with 'Login' text

![image](https://github.com/nicitaacom/22_aer/assets/39565703/2c7d2f49-dee2-4da7-8bed-1beccc2a7212)


Reproduce steps - login with google porvider
Now you able to see login when you already logged in

Alternatively you can write 'Logout' instead of 'Login' - its way simpler and let user understand that he is logged in
</details>




<details>
<summary>20. Use cookie-based auth</summary>

https://www.youtube.com/watch?v=w3LD0Z73vgU
https://www.youtube.com/watch?v=-7K6DRWfEGM

</details>




<br/>
<br/>
<hr/>

## What about chek yourself next time?
Please check your code for some issues - or possible improvements next time (of course you can't cover everything but .txt and comments and UX when login/register...)


### Please update your branch from 'development' next time

nicitaacom reviewed (changes requested):
# Roadmap (deprecated)

- [ ] 1 - Don't use negative conditions

- [ ] 2 - Simplify code

- [ ] 3 - Use supabase tables
 
- [x] 4 - Delete native comments everywhere

- [ ] 5 - Delete .txt files

- [ ] 6 - ~~Change file extention to .ts~~

- [ ] 7 - Fix eslint errors

- [ ] 8 - Don't use 'as string' with keys

- [ ] 9 - ~~Don't use negative conditions~~

- [ ] 10 - Use modern emails please

- [ ] 11 - ~~Change ctx to context~~

- [x] 12 - Use figma design (naming)

- [ ] 13 - Use figma design (icons)

- [ ] 14 - Please write much better 'Changes' next time

- [x] 15 - Use components instead of native button tags

- [ ] 16 - Decide which input validation do you use

- [ ] 17 - Improve UX (validation errors)

- [ ] 18 - Let user undestand somehow that he is logged in

- [ ] 19 - ~~Fix issue when I'm still see 'login' when I logged in~~

- [ ] 20 - Use cookie-based auth

- [ ] 21. Create the error page

- [ ] 22. Specify id in auth.ts

arifm6 committed:
PR49 Requested Changes 2

Improved Env type checking by updating env.mjs file, added new env for sendgrid email, updated prisma schema, improved variable names to match figma, replaced button with component button, deleted unused txt files,

arifm6 committed:
Simplified validation function implementation

Simplified implementation of validation functions to be easier to understand

arifm6 committed:
Updated session strategy

Updated DB Session strategy. updated oAuth provider logos.

arifm6 commented:
> # Requested changes_2
> 1. Don't use negative conditions
> 2. Simplify code
> 3. Use supabase tables
> 4. Delete native comments everywhere
> 5. Delete .txt files
> 6. Change file extention to .ts
> 7. Fix eslint errors
> 8. Don't use 'as string' with keys
> 9. Don't use negative conditions
> 10. Use modern emails please
> 11. Change ctx to context
> 12. Use figma design (naming)
> 13. Use figma design (icons)
> 14. Please write much better 'Changes' next time
> 15. Use components instead of native button tags
> 16. Decide which input validation do you use
> 17. Improve UX (validation errros)
> 18. Let user undestand somehow that he is logged in
> 19. Fix issue when I'm still see 'login' when I logged in
> 20. Use cookie-based auth
> 
> 
> ## What about chek yourself next time?
> Please check your code for some issues - or possible improvements next time (of course you can't cover everything but .txt and comments and UX when login/register...)
> 
> ### Please update your branch from 'development' next time

1. sometimes it is fine to use negative conditionals. The examples you provided are examples where using negative conditionals is good because I used it to check truthiness of a string.
2. Overall simplified.
3. its a prisma migration file. It keeps track of changes to the db schema. Read more here https://www.prisma.io/docs/concepts/components/prisma-migrate/migration-histories
4. deleted the comments that seemed useless. 
5. Deleted
6. the .mjs extension specifies modules as opposed to regular scripts. It should be .mjs. It is not necessary to convert these to .ts just for the sake of it. For more info: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules https://v8.dev/features/modules#mjs
7. discuss in discord as it looks like the lint rules have changed? github checks is fine on this issue for linting and my machine does not give these lint errors either.
8. fixed
9. read 1.
10. one step at a time. This PR already has too much stuff.
11. i prefer ctx. Its like using res instead of response
12. good catch, updated.
13. updated but your feedback gave no info
14. you say it tells you nothing but then the example you used gives some pretty big detail thats self explanatory.. The changes I provided were just a summary. 
15. good catch. updated
16. same as response to 10. One step at a time, there is another task open for input validation.
17. same as 16 and 10.
18. not specified to do in the task but will set up proper rerouting and ux in another task.
19. duplicate of 18
20. cookies are a piece of data that can be used for session managed. JWT can be stored in cookies. I switched to the database session strategy for a different reason. 

Overall good catches on many issues. Any other issues we can discuss in call. Also, the latest deployments are failing lint checks because I updated type checking on env variables and it seems you do not have the neccessary env variables set up.

nicitaacom commented:
### 1 - Don't use negative conditions
> sometimes it is fine to use negative conditionals. The examples you provided are examples where using negative conditionals is good because I used it to check truthiness of a string.

1.1 - It means you want ignore this or how do I understand this response?
1.2 -
> The examples you provided are examples where using negative conditionals is good

I didn't provide examples where using negative conditions is good (send me plesae reference where I write this)

<br/>
<br/>




### 2 - Simplify code

> Overall simplified.

No - if you want don't do this - just message about it but don't do it as is - I mean without quality - I mean its better to do your best instead of relating on 'magic' and smth like 'I keep it as is for now'

So decide you don't want do this or you do this in the best way you can (I know you can better - its effect like 'I already did a lot and again rewrite this code - I don't want')





<br/>
<br/>




### 3 - Use supabase tables

>its a prisma migration file. It keeps track of changes to the db schema

Ok but I don't understand why you unse prisma and supabase
It add complexity in code and requires x2 docs
You can do auth + sending emails using supabase only
(Ok its your code and you may leave it but if I would smb who comes to our project I would say https://www.youtube.com/watch?v=AsBHAZxflVs)

I see here 2 solutions:
1 - you leave it 'as is'
2 - you do it with supabase only

(please let me know what you choose)


<br/>
<br/>


### 5 - Delete .txt files
I'm still see .txt files 

![image](https://github.com/nicitaacom/22_aer/assets/39565703/79ecb5ce-2709-4e2b-93e2-d1fddc5a06d5)

![image](https://github.com/nicitaacom/22_aer/assets/39565703/e249a7d2-d988-46dc-ae01-f94f2745cd16)






<br/>
<br/>




###  6 - Change file extention to .ts
Ok its your code but I would use .ts
Also I don't see reson of reference to dev mozilla org (its like - what you want to say from this?)
I mean I also can send you some links and like 'undestand it as you want'




<br/>
<br/>





### 7 - Fix eslint errors
Now I got this

![image](https://github.com/nicitaacom/22_aer/assets/39565703/81ea4a6e-1ac4-4b3c-9031-30ebffbc2682)


Please provide .env in discord and delete comments in .env





<br/>
<br/>






### 8 - Don't use 'as string' with keys

![image](https://github.com/nicitaacom/22_aer/assets/39565703/81cc99e5-d29c-4f80-9ad0-619561f54768)

Delete 'as string' with keys






<br/>
<br/>



### 10 - Use modern emails please

>one step at a time. This PR already has too much stuff.

How understand this?







<br/>
<br/>






### 13 - Use figma design (icons)

> updated but your feedback gave no info

I will know that use figma design it not enogh
Whan info you want here? (so I can become better wiht my feedback)

I unable to check it because I haven't .env



<br/>
<br/>




### 14 - Please write much better 'Changes' next time

> you say it tells you nothing but then the example you used gives some pretty big detail thats self explanatory.. The changes I provided were just a summary.

Write problems that you solved - (of courst as developer I want know what you did in schema.prisma for example)
In feature write some problems that you solved for user (as I menioned you setted up validation (for example) and it helps user understand that he need to enter at lease 8 symbols in password (for example)) - but for user it doesn't matter taht you changed schema or created some hook or set up API

Reply to this so I can understand that you undestand what I mean and want to read from you in 'Changes' next time
(you may explain how you understood it)



<br/>
<br/>



### 16 - Decide which input validation do you use
>same as response to 10. One step at a time, there is another task open for input validation.

Please send link to this task


<br/>
<br/>



###  17 - Improve UX (validation errors)
>same as 16 and 10.

Please provide link


<br/>
<br/>



###  18 - Let user undestand somehow that he is logged in
> not specified to do in the task but will set up proper rerouting and ux in another task.

Change text in navbar from 'Login' to 'Logout' if user logged in

If you want to do this in separate taks - let me know you want me to assgin task for you for you want assing task for your self
(if you want asssign task for yourself - do it and message me in discrod about it becasue I may have no notification about new tasks in github) - so I can make sure that I up to date with new tasks in 22_aer






<br/>
<br/>

### 19 - Fix issue when I'm still see 'login' when I logged in
> duplicate of 18

Agreed







<br/>
<br/>




### 20 - Use cookie-based auth

>cookies are a piece of data that can be used for session managed. JWT can be stored in cookies. I switched to the database session strategy for a different reason.

I don't understand what do you want to say - also I don't understand reasons of use localStorage instead of cookies

<br/>
<br/>




<br/>
<hr/>



Also I don't understand why you check something as 'done' when its not done?
(if you don't want do it you can mark this ~~like so~~)
Also you can think that 'I did it' but when I check what you did it can be not exactly what I requested in my changes
(for example you marked  20 - Use cookie-based auth as 'done' but its not done - for other to-do as well)
So - please don't mark it as done because it can be misscomunicationg and I will not able to check is it done or you don't want do it or you done it but not as it should be or smth like that


<br/>


> Any other issues we can discuss in call.

Please specify thema and date&time GMT

nicitaacom reviewed (changes requested):
# Requested changes_3
<details>
<summary>1. Create the error page</summary>
In figma first - than mark it as ready to review (you may do it on github)
When figma design will approved do it in vs code

![image](https://github.com/nicitaacom/22_aer/assets/39565703/b3b10de9-f714-421c-85c3-7ef6dcce3ab8)

</details>




<details>
<summary>2.Specify id in auth.ts </summary>
You can just write 404 or smth like this (also you can use uuid)

![image](https://github.com/nicitaacom/22_aer/assets/39565703/055fe921-e4a2-4e4c-b1ff-32363f9576b1)

</details>




<details>
<summary>3.Improve 'Changes' </summary>

First of all change Notion task to github task
And adjust your changes taht suits user needs (I explained above)

</details>

nicitaacom commented:
# Changes

### 1. Roadmap moved to 'github project'
I did it to reduce mass and 'infinity' scroll on this page

arifm6 committed:
updated auth clarity

Updated navbar to reflect whether user is signed in or not. Made navbar login a nextAuth hook instead of a redirect. created consistent response returns (each response typically has a message and a status). Reverted nextAuth session strategy to jwt as nextAuth database session appears to not support credentials provider.
